cmd/piper.go

```diff
@@ -218,43 +221,100 @@ func checkTypes(config map[string]interface{}, options interface{}) map[string]i
-		case reflect.Bool:
-			// Sensible to convert strings "true"/"false" to respective boolean values as it is
-			// common practice to write booleans as string in yaml files.
-			paramValue = strings.ToLower(paramValue)
-			if paramValue == "true" {
-				config[paramName] = true
-				logWarning = false
-			} else if paramValue == "false" {
-				config[paramName] = false
-				logWarning = false
-			}
+			typeError = convertValueFromString(config, optionsField, paramName, paramValueType.String())
+		case reflect.Float32, reflect.Float64:
+			typeError = convertValueFromFloat(config, optionsField, paramName, paramValueType.Float())
+		case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+			typeError = convertValueFromInt(config, optionsField, paramName, paramValueType.Int())
+		default:
+			log.Entry().Warnf("Config value for '%s' is of unexpected type %s, expected %s. "+
+				"The value may be ignored as a result. To avoid any risk, specify this value with explicit type.",
+				paramName, paramValueType.Kind(), optionsField.Type.Kind())
 		}
 
-		if logWarning {
-			log.Entry().Warnf("Config value for '%s' is of unexpected type and is ignored", paramName)
+		if typeError != nil {
+			typeError = fmt.Errorf("config value for '%s' is of unexpected type %s, expected %s: %w",
+				paramName, paramValueType.Kind(), optionsField.Type.Kind(), typeError)
+			log.SetErrorCategory(log.ErrorConfiguration)
+			log.Entry().WithError(typeError).Fatal()
 		}
 	}
 	return config
 }
 
+func convertValueFromString(config map[string]interface{}, optionsField *reflect.StructField, paramName, paramValue string) error {
+	switch optionsField.Type.Kind() {
+	case reflect.Slice, reflect.Array:
+		// Could do automatic conversion for those types in theory,
+		// but that might obscure what really happens in error cases.
+		return fmt.Errorf("expected type to be a list (or slice, or array) but got string")
+	case reflect.Bool:
+		// Sensible to convert strings "true"/"false" to respective boolean values as it is
+		// common practice to write booleans as string in yaml files.
+		paramValue = strings.ToLower(paramValue)
+		if paramValue == "true" {
+			config[paramName] = true
+			return nil
+		} else if paramValue == "false" {
+			config[paramName] = false
+			return nil
+		}
+	}
+
+	return errIncompatibleTypes
+}
+
+func convertValueFromFloat(config map[string]interface{}, optionsField *reflect.StructField, paramName string, paramValue float64) error {
+	switch optionsField.Type.Kind() {
+	case reflect.String:
+		config[paramName] = strconv.FormatFloat(paramValue, 'f', -1, 64)
+		return nil
+	case reflect.Float32:
+		config[paramName] = float32(paramValue)
+		return nil
+	case reflect.Float64:
+		config[paramName] = paramValue
+		return nil
+	case reflect.Int:
+		if float64(int(paramValue)) == paramValue {
+			return nil
+		}
```

Comment:
No conversion?

Reply:
Maybe there should be. But let me explain... without this switch-case, this combination would be treated as type-mismatch error. But the json unmarshalling handles it in the condition checked by the if. So we return nil here just to suppress the type-mismatch error, but otherwise re-use whatever the json package does for type conversion.

Reply:
No, but you are right. It is silly not to also write the value into the config.

Reply:
resolved